A few notes for beginners on understanding the ecosystem (#21)
CHERIoT-Platform/CHERIoT-Platform.github.io: _posts/2024-11-08-navigating-the-ecosystem.markdown

---
layout: post
title:  "Finding your way around the CHERIoT ecosystem"

> date:   2024-11-07

Reviewer: This is yesterday.

categories: rtos sonata git
author: Phil Day
---
With a new set of folks in the DSbD cohort 6 just getting started with their Sonata boards I thought it might be useful to capture some of the things I've learn in the last few months about the ecosystem.

### Setting up a dev environment
The easiest and quickest way to get up and running is to use the dev container.
I have a fairly strong aversion to installing new software on my base system, and this just requires that you have docker installed.

> Opening a repo with dev container support in VS code and it will ask you if you want to reopen in a container.

Reviewer 2: Perhaps "Opening a repo with dev container support in VS code will ask you if you want to reopen in a container." or "Open a repo with dev container support in VS code and it will ask you if you want to reopen in a container."

> Do that and in a few seconds you're in a pre-build dev environment.

Reviewer 2: ```suggestion
Do that and in a few seconds you're in a pre-built dev environment.
```

Simples, to paraphrase a meerkat.

> One word of warning though; cheriot and Sonata both uses the ":latest" tag as the version of the docker image, so you won't automatically get updates when they are releases.

Reviewer 2: ```suggestion
One word of warning though; cheriot and Sonata both use the ":latest" tag as the version of the docker image, so you won't automatically get updates when they are released.
```

Should we be doing something differently?  Can we have the devcontainer poll for updates autonomically?

Reviewer: No, there's an open issue in VS Code to fix it, sadly.

If you've pulled an update to a repo and builds starts failing with no apparent reason it's worth pulling the latest dev container image and relaunching the dev container.
You can find the image name to pull in the .devcontainer/devcontainer.json file.

### Which Repo do I use ?

The CHERIoT ecosystem is split across a number of git repositories, and makes extensive used of [submodules](https://git-scm.com/book/en/v2/Git-Tools-Submodules) to link them and other external code from other places such as FreeRTOS.

> There is no version or release tagging in the repos, so you can only tell how current a submodule is by comparing the commit hashes.

Reviewer: Maybe clarify that this *will* happen after our 1.0 release and link to the 1.0 project?

On the plus side the CI system and maintainers do a good job of keeping it all compatible, and any breaking changes are discussed in the Signal channel.

#### [cheriot-rtos](https://github.com/CHERIoT-Platform/cheriot-rtos.git)
Cheriot-rtos is the core of the system, an includes its own set of examples that provide a great introduction to the main CHERI features.

> You don't even need any hardware to run them, as by default then build for the ibex-safe-simulator that's included in dev container.

Reviewer: Most of them default to building for the Sail model, the Ibex model has working input on the UART so a few use that.  You mostly don't need to notice this because `xmake run` does the right thing.

#### [cheriot-network](https://github.com/CHERIoT-Platform/network-stack.git)
This contains the code for the various compartments in the secure [network stack](https://cheriot.org/_rtos/networking/auditing/2024/03/08/cheriot-network-stack.html), and includes examples for SNTP, HTTP, HTTPS, and MQTT using some public servers.
Note though that this is really set up to be used as a submodule; there is no dev container support in here, and the examples will only build if this is cloned alongside cheriot-rtos.

It's worth looking at the layout of the code in here and how it is pulled into the other examples as an template of how to create reusable compartments.

#### [cheriot-demos](https://github.com/CHERIoT-Platform/cheriot-demos.git)
This contains examples that are more complete systems than the feature focused examples in cheriot-rtos.
It includes cheriot-rtos and cheriot-network as submodules, so you can also run all of the examples in those repos from here as well.
The commits of the submodule lags a little bit behind the main repos, but is generally recent enough unless you want to live on the bleeding edge.

> I think this is a better place for most folks to start, especially if you want at some point to use the network stack.

Reviewer 2: Would it make sense to start with this in the list and then drill down into the RTOS and network-stack submodules?

> You can still run all of the examples, and buffered a bit from changes in rtos, have access to some basic examples of working Sonata, and if you create an interesting demo your all set to contribute it back.   

Reviewer 2: ```suggestion
You can still run all of the examples, and buffered a bit from changes in rtos, have access to some basic examples of working Sonata, and if you create an interesting demo you're all set to contribute it back.
```

The current examples are:

> HughTheLightbulb requires a Sonata board connected via ethernet to network with a DHCP server and access to pool.ntp.org and test.mosquitto.org.

Reviewer 2: Perhaps make this a list, and maybe emphasize the name of the example?
```
- _HughTheLightbulb_ requires ...
```

It then controls the RGB LEDs on the Sonata Board via MQTT and shows the CPU and heap usage on the LCD.

> It's main focus is on the fault isolation within the [network stack](https://cheriot.org/_rtos/networking/auditing/2024/03/08/cheriot-network-stack.html), but it's also a useful example of basic IO programming on the Sonata board. 

Reviewer 2: ```suggestion
Its main focus is on the fault isolation within the [network stack](https://cheriot.org/_rtos/networking/auditing/2024/03/08/cheriot-network-stack.html), but it's also a useful example of basic IO programming on the Sonata board. 
```

> Compartmentalisation runs on Sonata or Arty A7 boards, and shows how compartments can be used to create an isolation boundary around a JavaScript runtime. It also shows how to expose the MMIO operations to JS.  There is also a version of this in cheriot-rtos/exercises that runs just in the ibex-safe-simulator. 

Reviewer: ```suggestion
Compartmentalisation runs on Sonata or Arty A7 boards, and shows how compartments can be used to create an isolation boundary around a JavaScript runtime.
It also shows how to expose the MMIO operations to JS.
There is also a version of this in cheriot-rtos/exercises that runs just in the ibex-safe-simulator. 
```

(No text change, one line per sentence)

---
layout: post
title:  "Finding your way around the CHERIoT ecosystem"
date:   2024-11-08
categories: rtos sonata git
author: Phil Day
---

> With a new set of folks in the DSbD cohort 6 just getting started with their Sonata boards I thought it might be useful to capture some of the things I've learn in the last few months about the ecosystem.

Reviewer 3: ```suggestion
With a new set of folks in the DSbD cohort 6 just getting started with their Sonata boards I thought it might be useful to capture some of the things I've learnt in the last few months about the ecosystem.
```

### Setting up a dev environment
The easiest and quickest way to get up and running is to use the dev container.
I have a fairly strong aversion to installing new software on my base system, and this just requires that you have docker installed.
Opening a repo with dev container support in VS code will ask you if you want to reopen in a container.
Do that and in a few seconds you're in a pre-built dev environment.
Simples, to paraphrase a meerkat.

> One word of warning though; cheriot and Sonata both use the ":latest" tag as the version of the docker image, so you won't automatically get updates when they are released.

Reviewer 3: I think you suggested we fix this but we never got round to it... Would be good to explicitly tie the `devcontainer.json` in the rtos with a known compatible dev container (semantic) version. Wouldn't that force a rebuild when necessary?

Author: There is now a date tagged version of the dev container, but we haven't updated any of the repos to use it yet.   I agree that would be worth doing (at least for cherito-demos, rtos might prefer to stay on the edge I guess).   Though for now best to just warn folks  to look out for it ;-)

Reviewer: It's normally fine to use a newer one.  We might want to enforce a minimum.  It would probably be annoying to keep bumping it for each new build and I don't want to make it too hard for people to use a newer version.

Reviewer: Actually, rereading @rmn30's suggestion, I think the following would work:

 - Modify the dev container deploy script to have a 'minimum-features-n' tag or similar that it pushes to at the same time as `latest`
 - Add this to the dev container json file.

This would let us roll out a new dev container with new features and then bump the RTOS to always depend on at least that version.  Suggestions for better names welcome.

If you've pulled an update to a repo and builds starts failing with no apparent reason it's worth pulling the latest dev container image and relaunching the dev container.
You can find the image name to pull in the .devcontainer/devcontainer.json file.

### Which Repo do I use ?

The CHERIoT ecosystem is split across a number of git repositories, and makes extensive used of [submodules](https://git-scm.com/book/en/v2/Git-Tools-Submodules) to link them and other external code from other places such as FreeRTOS.
There is no version or release tagging in the repos, so you can only tell how current a submodule is by comparing the commit hashes.

> Versioning will start with the CHERIoT 1.0 release, which you can track [here[(https://github.com/orgs/CHERIoT-Platform/projects/2).

Reviewer: ```suggestion
Versioning will start with the CHERIoT 1.0 release, which you can track [here](https://github.com/orgs/CHERIoT-Platform/projects/2).
```

On the plus side the CI system and maintainers do a good job of keeping it all compatible, and any breaking changes are discussed in the Signal channel.

#### [cheriot-rtos](https://github.com/CHERIoT-Platform/cheriot-rtos.git)

> Cheriot-rtos is the core of the system, an includes its own set of examples that provide a great introduction to the main CHERI features.

Reviewer 3: ```suggestion
Cheriot-RTOS is the core of the system, and includes its own set of examples that provide a great introduction to the main CHERI features.
```

> You don't even need any hardware to run them, as by default they build for one of the two emulator environments (sail and ibex-safe-simulator) that are included in dev container.

Reviewer 3: ```suggestion
You don't even need any hardware to run them, as by default they build for one of the two emulator environments (sail and ibex-safe-simulator) that are included in the dev container.
```

On the plus side the CI system and maintainers do a good job of keeping it all compatible, and any breaking changes are discussed in the Signal channel.

#### [cheriot-rtos](https://github.com/CHERIoT-Platform/cheriot-rtos.git)

> Cheriot-RTOS is the core of the system, and includes its own set of examples that provide a great introduction to the main CHERI features.

Reviewer: ```suggestion
CHERIoT-RTOS is the core of the system, and includes its own set of examples that provide a great introduction to the main CHERI features.
```
